File: 127WordLadder/wordladder.py
class Solution:
    def oneAway(self, firstWord, secondWord):
        firstDiffEncountered = False
        for i in range(len(firstWord)):
            if firstWord[i] != secondWord[i]:
                if firstDiffEncountered == True:
                    return False
                firstDiffEncountered = True

        return True

    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        if beginWord == endWord:
            return 1
        if endWord not in wordList:
            return 0

        answer = 1
        queue = [beginWord, None]

        while (len(queue) > 0):
            #print(queue)
            if queue[0] == None:
                queue = queue[1:]
                queue.append(None)
                answer +=1
                if queue[0] == None:
                    break
            for i in range(len(wordList)):
                if wordList[i] == beginWord:
                    wordList[i] = None
                elif wordList[i] != None and Solution.oneAway(self, queue[0], wordList[i]):
                    if wordList[i] == endWord:
                        return answer + 1
                    queue.append(wordList[i])
                    wordList[i] = None
                    #print(wordList)
            queue = queue[1:]

        return 0

sol = Solution
print(Solution.ladderLength(Solution, "hit", "cog", ["hot","dot","dog","lot","log","cog"]))

File: 127WordLadder/wordladder.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:

    bool oneEditAway(string firstWord, string secondWord){
        // all words have same length assumption given, so no check
        size_t wordSize = firstWord.size();
        bool firstDifferenceEncountered = false;
        for (int i = 0; i != wordSize; ++i){
            if (firstWord[i] != secondWord[i]){
                if (firstDifferenceEncountered){
                    return false;
                }
                firstDifferenceEncountered = true;
            }
        }
        return true;
    }
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if (find(wordList.begin(), wordList.end(), endWord) == wordList.end()){
            return 0;
        }
        size_t dictSize = wordList.size();
        bool visited[dictSize] = { false };
        
        queue<string> myqueue;
        myqueue.push(beginWord);
        myqueue.push("");

        int shortestSequence = 1;
        while(!myqueue.empty()){
            
            if (myqueue.front() == ""){
                myqueue.pop();
                ++shortestSequence;
                myqueue.push("");
                if (myqueue.front() == ""){
                    break;
                }
            }
            if (myqueue.front() == endWord){
                return shortestSequence;
            }
            
            for (int i = 0; i != dictSize; ++i){
                if (visited[i] == false && oneEditAway(myqueue.front(), wordList[i])){
                    myqueue.push(wordList[i]);
                    visited[i] = true;
                }
            }
 
            myqueue.pop();
        }
        return 0;
    }
};

int main(){
    Solution solution;
    vector<string> dictionary = {"hot","dot","dog","lot","log","cog", "fog", "fob"};

    cout << solution.ladderLength("hit", "fob", dictionary) << endl;

    return 0;
}